internal/server/routes.go

```diff
@@ -297,6 +304,33 @@ type isBlockingRequest interface {
 	IsBlockingRequest() bool
 }
 
+func requestVersion(req *http.Request) (*semver.Version, error) {
+	headerVer := req.Header.Get("Infra-Version")
+	if headerVer == "" {
+		return nil, fmt.Errorf("%w: Infra-Version header is required. The current version is %s", internal.ErrBadRequest, internal.FullVersion())
+	}
+	reqVer, err := semver.NewVersion(headerVer)
+	if err != nil {
+		return nil, fmt.Errorf("%w: invalid Infra-Version header: %v. Current version is %s", internal.ErrBadRequest, err, internal.FullVersion())
+	}
+	return reqVer, nil
+}
+
+func handlerForVersion(versions []routeVersion, reqVer *semver.Version) func(c *gin.Context) {
+	if reqVer == nil {
+		return nil
+	}
+
+	// TODO: iterate in reverse order to benefit newer versions over older versions
+	for _, v := range versions {
```

Comment:
I may have missed it, but I didn't see anywhere where the versions of the route are being sorted. This means someone someone could add a deprecated route in the wrong order in `addPreviousVersionHandlers()` where an older route could match first instead of the correct route.

e.g. You have handlers for:
* A = (0.0.0->0.15.0)
* B = [0.15.0->0.16.0)
* C = [0.16.0, latest)

There is an entry in `addPreviousVersionHandlers()` for the change from 0.15.0 -> 0.16.0 and one for 0.16.0 -> latest.

If the person writing the second handler accidentally put the 0.0.0 -> 0.15.0 handler in first, and a client called requesting 0.16.5, handler A could match instead of handler B since reqVer is greater. This isn't an issue now, but I would worry about maintainability in the future.

Reply:
Bruce asked this as well. I added a test `TestAPI_AddPreviousVersionHandlers_Order` in this PR to make sure the order is correct.

I think a test is nicer than sorting than because it ensures the order is correct in the source code. That way a reader can read it in the correct order, instead of people adding routes in bad orders and forcing every reader to sort it in their head.

Reply:
Could you put in a comment here which explains the ordering? I agree the tests should also cover this, but it should be clear to future authors that the order is important.